wgengine/netlog/logger.go

```diff
@@ -0,0 +1,235 @@
+		// the subnet range that Tailscale IP addresses are allocated from.
+		withinRoute = addrs[a]
+		for p := range prefixes {
+			if p.Contains(a) && p.Bits() > 0 {
+				withinRoute = true
+			}
+		}
+		return withinRoute && tsaddr.IsTailscaleIP(a), withinRoute && !tsaddr.IsTailscaleIP(a)
+	}
+
+	type tupleCounts struct {
+		flowtrack.Tuple
+		tunstats.Counts
+	}
+	var virtualTraffic, subnetTraffic, exitTraffic []tupleCounts
+	for conn, cnts := range stats {
+		srcIsTailscaleIP, srcWithinSubnet := classifyAddr(conn.Src.Addr())
+		dstIsTailscaleIP, dstWithinSubnet := classifyAddr(conn.Dst.Addr())
+		switch {
+		case srcIsTailscaleIP && dstIsTailscaleIP:
+			virtualTraffic = append(virtualTraffic, tupleCounts{conn, cnts})
+		case srcWithinSubnet || dstWithinSubnet:
+			subnetTraffic = append(subnetTraffic, tupleCounts{conn, cnts})
+		default:
+			const anonymize = true
+			if anonymize {
+				if len(exitTraffic) == 0 {
+					exitTraffic = []tupleCounts{{}}
+				}
+				exitTraffic[0].Counts = exitTraffic[0].Counts.Add(cnts)
+			} else {
+				exitTraffic = append(exitTraffic, tupleCounts{conn, cnts})
+			}
+		}
+	}
+
+	if b, err := json.Marshal(struct {
+		Start          time.Time     `json:"start"`
+		End            time.Time     `json:"end"`
+		VirtualTraffic []tupleCounts `json:"virtualTraffic,omitempty"`
+		SubnetTraffic  []tupleCounts `json:"subnetTraffic,omitempty"`
+		ExitTraffic    []tupleCounts `json:"exitTraffic,omitempty"`
+	}{start.UTC(), end.UTC(), virtualTraffic, subnetTraffic, exitTraffic}); err != nil {
+		logger.Logf("json.Marshal error: %v", err)
+	} else {
+		logger.Logf("%s", b)
+	}
+}
+
+func makeRouteMaps(routerCfg *router.Config) (addrs map[netip.Addr]bool, prefixes map[netip.Prefix]bool) {
+	addrs = make(map[netip.Addr]bool)
+	for _, p := range routerCfg.LocalAddrs {
+		if p.IsSingleIP() {
+			addrs[p.Addr()] = true
+		}
+	}
+	prefixes = make(map[netip.Prefix]bool)
+	for _, p := range routerCfg.Routes {
+		if p.IsSingleIP() {
+			addrs[p.Addr()] = true
+		} else {
+			prefixes[p] = true
+		}
+	}
+	return addrs, prefixes
+}
+
+// ReconfigRoutes configures the network logger with updated routes.
+func (nl *Logger) ReconfigRoutes(routerCfg *router.Config) {
+	nl.mu.Lock()
+	defer nl.mu.Unlock()
+	// TODO(joetsai): There is a race where deleted routes are not known at
+	// the time of extraction. We need to keep old routes around for a bit.
+	nl.addrs, nl.prefixes = makeRouteMaps(routerCfg)
+}
+
+// Shutdown shuts down the network logger.
+// This attempts to flush out all pending log messages.
+func (nl *Logger) Shutdown(ctx context.Context) error {
+	nl.mu.Lock()
+	defer nl.mu.Unlock()
+	if nl.logger == nil {
+		return nil
+	}
+	nl.cancel()
+	nl.mu.Unlock()
+	nl.group.Wait() // do not hold lock while waiting
+	nl.mu.Lock()
+	err := nl.logger.Shutdown(ctx)
+
+	nl.nodeID = logtail.PrivateID{}
+	nl.domainID = logtail.PrivateID{}
+	nl.logger = nil
+	nl.cancel = nil
```

Comment:
Is it worth `nil`ing out `nl.addrs` and `nl.prefixes` too?

Reply:
Done.